### ISSUE TITLE
Ability to set working directory

### ISSUE THREAD

**Issue by RPGHacker** [Fernap/UberASMTool#13]
It appears that UberASM Tool always looks for the `library` directory relative to its executable path. (This time I did make sure to check the manual more thoroughly, so I'm at least 90% convinced I didn't miss anything. 👀)

This can be a bit impractical when working with UberASM Tool as part of a build system that generates a directory structure for it, since it technically means that the executable itself (and all its dependencies) also need to be copied into that directory structure.

For this purpose, it would be greate if the location of the library directory could be overridden from a parameter in `list.txt`, in a similar manner to how the `macrolib` file path can be overridden.

I'm pretty sure the same thing goes for the `routines` directory, which I assume will have a similar problem (I don't have any resource that uses routines yet, so I can't verify).

Simple examples:

```
global:         other/global_code.asm
statusbar:      other/status_code.asm
macrolib:       other/macro_library.asm
freeram:        $7FAC80                    ; 2 bytes of free RAM
library:        library
routines:       routines
```

(For backwards-compatibility, these could be made optional)

**Comment by Fernap** [comment]
Yeah, this is something that I'd already kind of wanted to add, but didn't quite get to.  I had been thinking about doing this as a command line option instead, but having it the list file might be a little easier.  This might be a bit tricky though with the order that things happen in the list file processing.  I'll figure something out here though.

**Comment by RPGHacker** [comment]
For what it's worth, at least for my purposes, command line option and list file would both work. I mainly suggested the list file because it seems a little more consistent and because I assume it'll be a bit easier for the average user to work with.

A third potential solution could be to make the library path relative to list.txt instead of the executable (though I assume it'll have the same ordering problem you've mentioned above, and it might also break backwards-compatibility).

**Comment by Fernap** [comment]
If you're still watching this, RPGHacker, would you mind describing your intended use case in a little more detail?  At first I thought this was to keep a single instance of the executable, while allowing it to run for different projects, which would necessitate separate copies of library and routines, like you mentioned...but if this were the case, you'd likely also need to be able to specify different resource directories (level/, etc).

Or maybe you were thinking that you had multiple projects that you wanted to share libraries for, while separate instances of the executable were otherwise inconsequential?

I think the latter could be achieved by using a symlink/junction, without any further changes.

**Comment by RPGHacker** [comment]
Certainly!

My use case is that I built my own little SMW build system in Python for the primary purpose of being able to easily test my own patches. Basically, I can create different ROMs from scratch with different settings by setting some parameters in a helper GUI.

Now a lot of times, certain bugs will occur in my patches only when used in conjunction with very specific resources, so to make testing those interactions easier, I also add compatibility-testing options to my patch configurations. By default, they would be set to "none" to only apply my own patches, but you could (for example) change them to "custom power-up patch" to additionally patch that to the ROM. This is a very practical way of working once set up.

Since these compatibility settings need to be configurable completely via the GUI and there can be infinite permutations, it means I can't use pre-existing list files or directory structures to apply these patches. Instead, I define how to apply a patch in my Python configuration, and my build system then generates the list files and directory structures as needed into some temp output directory.

What I do with most tools is that I generate a list file in the temp directory and then copy all resources and dependencies into that same directory. With UberASMTool, I'm attempting to do the same thing, but can only do so with some complications, due to the fact it always looks for the library directory next to the executable. If a resource comes bundled with some library files, normally my approach would be to copy those library files into the output diretory tree along with the resource. However, currently that is only possible if I also copy the executable itself (and all its dependencies) into the output directory and run it from there (which is certainly possible, and I might be doing that right now, but it does add one or two complications to my code). The only other alternative I can think of would be to make my build system copy the library files into the library directory that already exists next to the executable, however, I don't want to do that, since I consider all outputs of my build system to be temporary files that should never be commited to source control, whereas all the tools are meant to be commited. Doing this would kinda pollute the UberASMTool base directory. Also, it might create problems over time, since patches might at some point gain access to library files they were never meant to work with.

From my end, the simplest solution I can think of would be if I could just override the library directory via the list file or a command line argument. In that case, I could just point it to my output directory and wouldn't need to bother with copying the executable. (IIRC) I don't have any problems with resource directories, since those can already be repointed via (I think) command line arguments, so those can easily use my auto-generated output directory.

**Comment by Fernap** [comment]
I see, thanks...and yeah, there's also no way to do this for resource directories (level, etc).  That one can be worked around separately because you can always just give an absolute path to the resource .asm file, instead of having of having UAT look in the level/ folder next to the executable.

Would it work just as well if you could simply have UAT change its working directory via a command line option?  So if I'm understanding what you said, you'd run the tool with this option, and set the working directory to your temp directory, which would contain library/, routines/, level/, etc.

**Comment by RPGHacker** [comment]
I think that should work, yeah! I sort of used a similar workaround for GPS already, which (most likely unintentionally) uses the current working directory when accessing certain directories.

**Comment by RPGHacker** [comment]
Thanks for implementing the new setting so quickly! I was only able to give it a try just now. Unfortunately, it does turn out to not fully solve my specific use case.

```
error: (Efile_not_found): File '[path_to_output_directory]\uberasm\asm/base/clean.asm' wasn't found.
```

Because I repointed UberASM's working directory to my generated output directory with the new `-d` argument, it now also expects to find all of its dependencies in there. I could of course solve this issue by copying UberASM's `asm` directory to my generated output directory as well, but since my original goal was to have all the user-defined data separated from UberASM's static dependencies, this does still feel somewhat unsatisfactory.

It appears to me that a `-l` or `--library-dir` command line argument (for overwriting only the library directory path) would still be required to solve my specific use case.

**Comment by Fernap** [comment]
Oh no...I'll figure something out, but it may be a little bit...it might just be as simple as including the original executable folder in asar's path when patching stuff.
